libs/dyn/mapping.go

```diff
@@ -46,9 +47,15 @@ func newMappingFromGoMap(vin map[string]Value) Mapping {
 	return m
 }
 
-// Pairs returns all the key-value pairs in the Mapping.
+// Pairs returns all the key-value pairs in the Mapping. The pairs are sorted by
+// their key in lexicographic order.
 func (m Mapping) Pairs() []Pair {
-	return m.pairs
+	pairs := make([]Pair, len(m.pairs))
+	copy(pairs, m.pairs)
+	sort.Slice(pairs, func(i, j int) bool {
```

Comment:
Why do we need it?

Reply:
To make the order in which `dyn.MapByPattern` walks configuration fields deterministic. 

Even though `dyn.Mapping` represents key values fields as a slice, the order in which elements in the slice are present are influenced by multiple sources of randomness like the order in which configuration files are parsed, glob patterns are expanded or empty values are added to the configuration tree during normalization.

Without this we wont be able to make the assertions on `[]dyn.Location` and `[]dyn.Path` we make in the unit tests added in this PR.

This modification also makes `Pairs()` function nicer since the order now is guaranteed based on the content, similar to how maps work in C++.

Reply:
> similar to how maps work in C++.

Except only they implemented differently in C++ and not with sorting :)

> Without this we wont be able to make the assertions on []dyn.Location and []dyn.Path we make in the unit tests added in this PR.

Instead of `assert.Equal(t, tc.diagnostics, diags)` you could loop through paths / locations and do the Contains, right?

The downside is that what used to be a O(1) call becomes O(nlogn) call now and we use .Pairs() extensively (think as an example of `visit` call) 

> randomness like the order in which configuration files are parsed, glob patterns are expanded or empty values are added to the configuration tree during normalization

Is it really true? Most if not all of this properties are represented as slices and should be deterministic. 

If we really need alphabetical order and save the performance we could do the sorting on the `Add / Append` to the map operation

Reply:
> Is it really true? Most if not all of this properties are represented as slices and should be deterministic.

Yeah, examples of non-determinisim include:
1. https://github.com/databricks/cli/blob/383d580917a8459eeebd1ce54d95e65afaa13aa7/libs/dyn/convert/normalize.go#L128
2. https://github.com/databricks/cli/blob/383d580917a8459eeebd1ce54d95e65afaa13aa7/libs/dyn/mapping.go#L43

The feedback about performance is fair. I figured it dwarfs the API calls/file IO but let me think of a way to retain performance.

Reply:
Done, now we sort the locations inline to assert the diags.